### PR TITLE
Caps flockdrones at 75 per flock

### PR DESCRIPTION
**Pull request by TobleroneSwordfish**
[BALANCE] [GAMEMODES]
## About the PR 
See above, attempting to make a new drone when above the cap displays `Flock complexity too high, unable to support additional drones.` and cages will just spit out some coagulated gnesis onto the floor.
Also cleans up some related egg cost code to reduce copy paste.

## Why's this needed? 
- The mob AI loop kinda stops working past a certain point leading to hundreds of near frozen drones running at extremely low tickrate.
- Discourages flock from just spamming more and more drones, forcing them to build more structures and complete the relay.

## Changelog

```changelog
(u)LeahTheTech
(*)Flock are now capped at 75 flockdrones per flock.
```

### CODE REVIEW

File: code/mob/living/critter/ai/flock/flocktasks.dm [hunk 1]
```diff
@@ -136,10 +136,10 @@ stare
 	add_task(holder.get_instance(/datum/aiTask/succeedable/replicate, list(holder)))
 
 /datum/aiTask/sequence/goalbased/flock/replicate/precondition()
-	. = FALSE
 	var/mob/living/critter/flock/drone/F = holder.owner
-	if(F?.can_afford(FLOCK_LAY_EGG_COST + clamp((F.flock.getComplexDroneCount() - FLOCK_MIN_DESIRED_POP) * FLOCK_ADDITIONAL_RESOURCE_RESERVATION_PER_DRONE, 0, FLOCK_LAY_EGG_COST * 2)))
-		. = TRUE
+	if (!F.flock)
```
Comment (frawhst) [inline code/mob/living/critter/ai/flock/flocktasks.dm:140]: Some weird nullchecking going on here- line 140 will runtime if `F == null`, but you still nullcheck it on line 142. Similarly we return on 141 if `F.flock == null`, but `F.flock` still gets null checked below.

tldr `if (!F?.flock)` and remove the others

Comment (frawhst) [reply code/mob/living/critter/ai/flock/flocktasks.dm:140]: similar thing on line 200

Comment (TobleroneSwordfish) [reply code/mob/living/critter/ai/flock/flocktasks.dm]: frankencode

**Comment by BatElite** [comment]
Can they sort of get around this cap by repeatedly breaking new drones into flockbits? thinking more in terms of mob overloading than balance

File: code/obj/flock/cage.dm [hunk 2]
```diff
@@ -164,8 +164,12 @@
 	process()
 		// process fluids into stuff
 		if(reagents.has_reagent(target_fluid, create_egg_at_fluid))
-			reagents.remove_reagent(target_fluid, create_egg_at_fluid)
-			spawnEgg()
+			if (src.flock?.getComplexDroneCount() < FLOCK_DRONE_LIMIT)
+				spawnEgg()
+			else
+				src.reagents.reaction(get_turf(src), react_volume = create_egg_at_fluid/2)
```
Comment (amylizzle) [inline code/obj/flock/cage.dm:170]: maybe resource cubes instead of flock juice?

**Review by amylizzle — APPROVED** [review]
Looks good. Minor change suggested.

File: code/obj/flock/cage.dm (hunk 2 above)
Comment (FlameArrow57) [reply code/obj/flock/cage.dm:170]: Yeah resources cubes would be cool

**Comment by FlameArrow57** [comment]
Going with what BatElite said, think a Flockbit max count might be needed to otherwise you could get around this.

Also 75 drones * 10 compute per is 750 compute.

File: code/datums/flock/flock.dm [hunk 3]
```diff
@@ -495,6 +495,9 @@ var/flock_signal_unleashed = FALSE
 		return 0
 	return length(src.units[/mob/living/critter/flock/drone/])
 
+/datum/flock/proc/currentEggCost()
+	return FLOCK_CONVERT_COST + FLOCK_LAY_EGG_COST + clamp((src.getComplexDroneCount() - FLOCK_MIN_DESIRED_POP) * FLOCK_ADDITIONAL_RESOURCE_RESERVATION_PER_DRONE, 0, FLOCK_LAY_EGG_COST * 2)
```
[the remark this answers is not in the document]
Comment (TobleroneSwordfish) [reply code/datums/flock/flock.dm:499]: ohh you're right, I thought it was odd that a lot of stuff didn't update to the new cost, will fix

**Comment by TobleroneSwordfish** [comment]
I'm not too worried about the flockbit spam since they are at least hopefully a lot lighter performance wise (but there's 3 of them per sooo 🤷)
As for the number, it's hard to gauge how many drones exist on a map at once during a live round without admin tools so it may well be too high but can be easily adjusted.

File: code/mob/living/critter/ai/flock/flocktasks.dm [hunk 4]
```diff
@@ -94,6 +94,14 @@ stare
 	src.subtasks = list() //get rid of the move and replace it with flockmove
 	add_task(holder.get_instance(/datum/aiTask/succeedable/move/flock, list(holder)))
 
+///The amount of resources a drone needs to be eligible to lay an egg (eggs still only cost FLOCK_LAY_EGG_COST)
+/datum/aiTask/sequence/goalbased/flock/proc/current_egg_cost()
+	var/mob/living/critter/flock/flockcritter = src.holder.owner
+	if (!flockcritter?.flock)
+		return 100
```
Comment (FlameArrow57) [inline code/mob/living/critter/ai/flock/flocktasks.dm:101]:
```suggestion
		return FLOCK_LAY_EGG_COST
```

File: code/mob/living/critter/ai/flock/flocktasks.dm [hunk 5]
```diff
@@ -94,6 +94,14 @@ stare
...
+	return FLOCK_CONVERT_COST + FLOCK_LAY_EGG_COST + clamp((flockcritter.flock.getComplexDroneCount() - FLOCK_MIN_DESIRED_POP) * FLOCK_ADDITIONAL_RESOURCE_RESERVATION_PER_DRONE, 0, FLOCK_LAY_EGG_COST * 2)
```
Comment (FlameArrow57) [inline code/mob/living/critter/ai/flock/flocktasks.dm:102]:
```suggestion
	return FLOCK_LAY_EGG_COST + clamp((flockcritter.flock.getComplexDroneCount() - FLOCK_MIN_DESIRED_POP) * FLOCK_ADDITIONAL_RESOURCE_RESERVATION_PER_DRONE, 0, FLOCK_LAY_EGG_COST * 2)
```

File: code/mob/living/critter/ai/flock/flocktasks.dm [hunk 6]
```diff
@@ -197,7 +205,9 @@ stare
 /datum/aiTask/sequence/goalbased/flock/nest/precondition()
 	. = FALSE
 	var/mob/living/critter/flock/drone/F = holder.owner
-	if(F?.can_afford(FLOCK_CONVERT_COST + FLOCK_LAY_EGG_COST + clamp((F.flock.getComplexDroneCount() - FLOCK_MIN_DESIRED_POP) * FLOCK_ADDITIONAL_RESOURCE_RESERVATION_PER_DRONE, 0, FLOCK_LAY_EGG_COST * 2)))
+	if (!F?.flock)
+		return
+	if(F.can_afford(src.current_egg_cost()) && F.flock.getComplexDroneCount() < FLOCK_DRONE_LIMIT)
```
Comment (FlameArrow57) [inline code/mob/living/critter/ai/flock/flocktasks.dm:210]:
```suggestion
	if(F.can_afford(FLOCK_CONVERT_COST + src.current_egg_cost()) && F.flock.getComplexDroneCount() < FLOCK_DRONE_LIMIT)
```

File: code/obj/flock/cage.dm [hunk 7]
```diff
@@ -164,8 +164,13 @@
 	process()
 		// process fluids into stuff
 		if(reagents.has_reagent(target_fluid, create_egg_at_fluid))
+			if (src.flock?.getComplexDroneCount() < FLOCK_DRONE_LIMIT)
+				spawnEgg()
+			else
+				var/obj/item/flockcache/cube = new(get_turf(src))
+				cube.resources = create_egg_at_fluid
```
Comment (FlameArrow57) [inline code/obj/flock/cage.dm:171]: I think this could result in several hundred resources produced per person caged?

Comment (TobleroneSwordfish) [reply code/obj/flock/cage.dm:171]: Yep, but it's the same ratio as would have been converted into drones so I think it's okay? Also bear in mind this only happens when flock is already massive and needs to be building structures anyway.

**Review by frawhst — APPROVED** [review]
